Mathlib/Data/Nat/Fib/Zeckendorf.lean

```diff
@@ -0,0 +1,177 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Nat.Fib.Basic
+
+/-!
+# Zeckendorf's Theorem
+
+This file proves Zeckendorf's theorem: Every natural number can be written uniquely as a sum of
+distinct non-consecutive Fibonacci numbers.
+
+## Main declarations
+
+* `List.IsZeckendorfRep`: Predicate for a list to be an increasing sequence of non-consecutive
+  natural numbers greater than or equal to `2`, namely a Zeckendorf representation.
+* `Nat.greatestFib`: Index of the greatest Fibonacci number less than or equal to some natural.
+* `Nat.zeckendorf`: Send a natural number to its Zeckendorf representation.
+* `Nat.zeckendorfEquiv`: Zeckendorf's theorem, in the form of an equivalence between natural numbers
+  and Zeckendorf representations.
+
+## TODO
+
+We could prove that the order induced by `zeckendorfEquiv` on Zeckendorf representations is exactly
+the lexicographic order.
+
+## Tags
+
+fibonacci, zeckendorf, digit
+-/
+
+open List Nat
+open scoped BigOperators
+
+-- TODO: The `local` attribute makes this not considered as an instance by linters
+@[nolint defLemma docBlame]
+local instance : IsTrans ℕ fun a b ↦ b + 2 ≤ a where
+  trans _a _b _c hba hcb := hcb.trans $ le_self_add.trans hba
+
+namespace List
+
+/-- A list of natural numbers is a Zeckendorf representation (of a natural number) if it is an
+increasing sequence of non-consecutive numbers greater than or equal to `2`.
+
+This is relevant for Zeckendorf's theorem, since if we write a natural `n` as a sum of Fibonacci
+numbers `(l.map fib).sum`, `IsZeckendorfRep l` exactly means that we can't simplify any expression
+of the form `fib n + fib (n + 1) = fib (n + 2)`, `fib 1 = fib 2` or `fib 0 = 0` in the sum. -/
+def IsZeckendorfRep (l : List ℕ) : Prop := (l ++ [0]).Chain' (fun a b ↦ b + 2 ≤ a)
+
+lemma IsZeckendorfRep.sum_fib_lt : ∀ {n l}, IsZeckendorfRep l → (∀ a ∈ (l ++ [0]).head?, a < n) →
+  (l.map fib).sum < fib n
+| n, [], _, hn => fib_pos.2 $ hn _ rfl
+| n, a :: l, hl, hn => by
+  simp only [IsZeckendorfRep, cons_append, chain'_iff_pairwise, pairwise_cons] at hl
+  have : ∀ b, b ∈ head? (l ++ [0]) → b < a - 1 :=
+    fun b hb ↦ lt_tsub_iff_right.2 $ hl.1 _ $ mem_of_mem_head? hb
+  simp only [mem_append, mem_singleton, ←chain'_iff_pairwise, or_imp, forall_and, forall_eq,
+    zero_add] at hl
+  simp only [map, List.sum_cons]
+  refine (add_lt_add_left (sum_fib_lt hl.2 this) _).trans_le ?_
+  rw [add_comm, ←fib_add_one (hl.1.2.trans_lt' zero_lt_two).ne']
+  exact fib_mono (hn _ rfl)
+
+end List
+
+namespace Nat
+variable {l : List ℕ} {a m n : ℕ}
+
+/-- The index of the greatest Fibonacci number less than or equal to `n`. -/
```

Comment:
This docstring doesn't specify how the tie is broken for `1`

Reply:
Better now?